src/models/krate.rs

```diff
@@ -101,50 +101,39 @@ pub struct NewCrate<'a> {
 }
 
 impl<'a> NewCrate<'a> {
-    pub fn create_or_update(self, conn: &mut PgConnection, uploader: i32) -> AppResult<Crate> {
+    pub fn update(&self, conn: &mut PgConnection) -> QueryResult<Crate> {
         use diesel::update;
 
-        conn.transaction(|conn| {
-            // To avoid race conditions, we try to insert
-            // first so we know whether to add an owner
-            if let Some(krate) = self.save_new_crate(conn, uploader)? {
-                return Ok(krate);
-            }
-
-            update(crates::table)
-                .filter(canon_crate_name(crates::name).eq(canon_crate_name(self.name)))
-                .set(&self)
-                .returning(Crate::as_returning())
-                .get_result(conn)
-                .map_err(Into::into)
-        })
+        update(crates::table)
+            .filter(canon_crate_name(crates::name).eq(canon_crate_name(self.name)))
+            .set(self)
+            .returning(Crate::as_returning())
+            .get_result(conn)
     }
 
-    fn save_new_crate(&self, conn: &mut PgConnection, user_id: i32) -> QueryResult<Option<Crate>> {
+    pub fn create(&self, conn: &mut PgConnection, user_id: i32) -> QueryResult<Crate> {
         use crate::schema::crates::dsl::*;
 
         conn.transaction(|conn| {
-            let maybe_inserted: Option<Crate> = diesel::insert_into(crates)
+            let krate: Crate = diesel::insert_into(crates)
                 .values(self)
                 .on_conflict_do_nothing()
```

Comment:
Do we still want `ON CONFLICT DO NOTHING` here if this is meant to be a straight insert of a new record?

Specifically, I'm wondering how this intersects with the crate owner code below now being unconditional. What happens if `create()` is called and the crate already exists?

Reply:
- if we use `on_conflict_do_nothing()`:
  - the `get_result()` will return `Ok(Crate)` if the crate was inserted into the database or `Err(NotFound)` if it already exists
  - the `?` after `get_result()` will ensure that the crate owner is not inserted in case of `Err(NotFound)`
  - calling `.optional()` on the return value of `create()` it transforms it into `Result<Option<Crate>>`

- if we **don't** use `on_conflict_do_nothing()`:
  - the `get_result()` will return `Ok(Crate)` if the crate was inserted into the database or `Err(DatabaseError)` if it already exists
  - the error case here would be a bit harder to handle since it will be represented as a somewhat random uniqueness constraint error

> the crate owner code below now being unconditional

it is still conditional through the early return caused by the `?` operator :)

Reply:
Gotcha. OK, I was mistaken about what `ON CONFLICT DO NOTHING` would do from a Rust perspective there.

Thanks!

Reply:
no worries, I've spent a significant amount of time today figuring this out 😅